types/react/index.d.ts

```diff
@@ -51,11 +51,15 @@ declare namespace React {
     // React Elements
     // ----------------------------------------------------------------------
 
-    type ReactType<P = any> =
+    type ElementType<P = any> =
         {
             [K in keyof JSX.IntrinsicElements]: P extends JSX.IntrinsicElements[K] ? K : never
         }[keyof JSX.IntrinsicElements] |
         ComponentType<P>;
```

Comment:
This looks like an opportunity to change from `ComponentType` to `JSXElementConstructor` here. If I understood my computer science correctly it'd make `P` contravariant instead of invariant. (Ideally for HOCs it'd be nice to have it be covariant instead, but no way to do it in typescript?)

```ts
function hasClassNameProp<T extends React.ReactType<{ className: string }>>(
    Component: T
) {
    return Component;
}

// these should all work (or only fail for excess property reasons)
// (all are components capable of receiving a { className: string; children: React.ReactNode })
hasClassNameProp('div');
hasClassNameProp((props: {}) => null);
hasClassNameProp((props: { className?: string }) => null);
hasClassNameProp((props: { className: string }) => null);
hasClassNameProp((props: { className: string; id?: string }) => null);
hasClassNameProp({} as any as React.FC<{}>);
hasClassNameProp({} as any as React.FC<{ className?: string }>);
hasClassNameProp({} as any as React.FC<{ className: string }>);
hasClassNameProp({} as any as React.FC<{ className: string; id?: string }>);
```

Reply:
I guess the root issue is that `ComponentType` uses `FunctionComponent` which has implicit `children`. Switching to `JSXElementConstructor` would get rid of this as far as I can tell. From my experience this is a good thing because we usually infer some union type if we use `ComponentProps` on a `ReactType<P>`.

Not sure what other contributors think about this change. I fear that a discussion about your suggested change might block the original change of this PR from being merged. Can we discuss this on a separate PR?

Reply:
@Jessidhia one way I was able to get P to be covariant is to do this:
```
function myHoc<P extends BaseProps, C>(Component: React.JSXElementConstructor<P> & C) {
  return Component
}
```

Allows for P being covariantly constraint-checked and C is inferred to grab static properties.

Reply:
@eps1lon I'm fine with the change as-is. We need to eventually remove implicit `children` out of everything anyways, and we need to educate people more before making a breaking change like this on which type to use as parameter types (contravariant) vs which type to use as annotated types (covariant)

Reply:
@ferdaber Alright. I'll definitely need to  go over those *-variant terms again. I'm not very confident with this terminology.

Could you explicitly approve the changes so that it's obvious for maintainers that CI is at fault here?

Reply:
This one is actually not about the implicit children but the implicit `.propTypes` and `.defaultProps` -- they're the ones currently making `ReactType` invariant.